### DISCUSSION TITLE
rust-analyzer reporting `file not included in module tree (unlinked-file)` if crate directory is inside another crate

### DISCUSSION THREAD

**Discussion by ttencate** [helix-editor/helix#6758] [Q&A]
### Summary

When a crate is a subdirectory of another crate, rust-analyzer fails to work and gives the error message `file not included in module tree (unlinked-file)` with a green squiggly line on the first three characters of the file.

I'm filing this as a bug with Helix, rather than with rust-analyzer, because VS Code has no such trouble. If I'm doing something wrong, please let me know.

### Reproduction Steps

```
$ cargo init foo
$ cd foo
$ cargo init bar
$ cd bar
$ helix src/main.rs 
```

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2023-04-14T15:01:31.620 helix_vcs [ERROR] PeelToCommit(
    Unborn {
        name: FullName(
            "refs/heads/main",
        ),
    },
)
2023-04-14T15:01:31.621 helix_vcs [ERROR] failed to open diff base for /tmp/foo/bar/src/main.rs
2023-04-14T15:01:31.621 helix_vcs [ERROR] PeelToCommit(
    Unborn {
        name: FullName(
            "refs/heads/main",
        ),
    },
)
2023-04-14T15:01:31.621 helix_vcs [ERROR] failed to obtain current head name for /tmp/foo/bar/src/main.rs
```

</details>

### Platform

Linux

### Terminal Emulator

n/a

### Helix Version

23.03

### rust-analyzer Version

1.68.2 (9eb3afe 2023-03-27)

**Comment by pascalkuthe** [comment]
rust analyzer is automatically started at the topmost `Cargo.toml` file in the workspace. To make the nested crate show up you can either use a cargo workspace (which is usually better) but if you really want separate nested crates then you can create a `.helix/config.toml` file at the repo root (same directory as `.git`) and manually configure the additional roots:

``` toml
[editor]
workspace-lsp-roots = ["bar"]
```

**Reply by ttencate** [reply]
I don't think there is a workspace here at all, at least not according to cargo, since there is no `[workspace]` section anywhere: https://doc.rust-lang.org/cargo/reference/workspaces.html. If I issue `cargo build` in `foo`, I don't see `bar` being built, or vice versa. `cargo build -p bar` does not work inside `foo` and vice versa. So it would seem that Helix has a different notion of "workspace root" than cargo does.

**Comment by antoyo** [comment]
Is there any reason why this doesn't work while opening 2 projects that are at the same level (e.g. not nested) work?
Since the non-nested case works, wouldn't it be easy to make the nested case work?

**Reply by the-mikedavis** [reply]
When they are not nested we can detect the roots accurately - the topmost `Cargo.toml`. When they're nested though the higher `Cargo.toml` is automatically detected as the workspace root for both which is incorrect for the nested project. Using `editor.workspace-lsp-roots` is the proper way to handle this
